File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        connect(root, null);
        return root;
    }
    public void connect(Node root, Node nextRight) {
        if (root == null) return;
        root.next = nextRight;
        if (root.left == null && root.right == null) {
            return;
        }
        connect(root.left, root.right);
        connect(root.right, nextRight == null ? nextRight: nextRight.left);
    }
}

File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        def sub_connect(curr_node, next_node):
            if curr_node == None:
                return
            curr_node.next = next_node
            if curr_node.left == None and curr_node.right == None:
                return
            sub_connect(curr_node.left, curr_node.right)
            sub_connect(curr_node.right, next_node if next_node is None else next_node.left)
        sub_connect(root, None)
        return root